src/lib.rs

```diff
@@ -124,6 +246,317 @@ pub enum DumpError<T> {
 //
 pub type DumpLzss = lzss::Lzss<6, 4, 0x20, { 1 << 6 }, { 2 << 6 }>;
 
+pub fn from_mem(addr: u32, buf: &mut [u8], _meta: bool) -> Result<(), ()> {
+    let src =
+        unsafe { core::slice::from_raw_parts(addr as *const u8, buf.len()) };
+
+    buf.copy_from_slice(src);
+    Ok(())
+}
+
+pub fn to_mem(addr: u32, buf: &[u8]) -> Result<(), ()> {
+    let dest =
+        unsafe { core::slice::from_raw_parts_mut(addr as *mut u8, buf.len()) };
+
+    // dest.copy_from_slice(buf); doesn't work here?!
+    for i in 0..buf.len() {
+        dest[i] = buf[i];
+    }
+
+    Ok(())
+}
+
+///
+/// Initialize the dump areas based on the specified list.
+///
+pub fn initialize_dump_areas(areas: &[DumpArea], chunksize: Option<usize>) -> Option<u32> {
+    let mut next = 0;
+
+    for area in areas.iter().rev() {
+        let chunksize = match chunksize {
+            Some(chunksize) => chunksize,
+            None => area.length as usize,
+        };
+
+        let length = area.length.min(chunksize as u32);
+
+        for offset in (0..area.length).step_by(chunksize).rev() {
+            let address = area.address + offset;
+
+            unsafe {
+                let header = address as *mut DumpAreaHeader;
+
+                //
+                // We initialize our dump headers with deliberately bad magic
+                // to prevent any dumps until we have everything initialized
+                //
+                (*header) = DumpAreaHeader {
+                    magic: DUMP_UNINITIALIZED,
+                    address: address,
+                    nsegments: 0,
+                    written: core::mem::size_of::<DumpAreaHeader>() as u32,
+                    length: length,
+                    agent: area.agent.into(),
+                    dumper: DUMPER_NONE,
+                    next,
+                }
+            }
+
+            next = address;
+        }
+    }
+
+    if areas.len() > 0 {
+        let mut address = areas[0].address;
+
+        while address != 0 {
+            unsafe {
+                let header = address as *mut DumpAreaHeader;
+                (*header).magic = DUMP_MAGIC;
+                address = (*header).next;
+            }
+        }
+
+        Some(areas[0].address)
+    } else {
+        None
+    }
+}
+
+///
+/// This should only be called in the context of the dump agent proxy.
+///
+pub fn get_dump_area<T>(
+    base: u32,
+    index: u8,
+    mut read: impl FnMut(u32, &mut [u8], bool) -> Result<(), T>,
+) -> Result<DumpArea, DumpError<T>> {
+    let mut address = base;
+    let mut i = 0;
+    const HEADER_SIZE: usize = core::mem::size_of::<DumpAreaHeader>();
+
+    loop {
```

Comment:
This loop looks [accidentally quadratic](https://accidentallyquadratic.tumblr.com/) to me, though I'm not sure if N ever gets large enough for it to matter.

Reply:
Added a comment. We're walking a linked list here, so it's going to be O(N).